### PR TITLE
speeding up sdxl and fixes

### PR DESCRIPTION
**Pull request by HDCharles**
Summary: removing epilogue fusion significantly speeds up dynamic
quantization, we get ~10% speedup. Also converted conv_1x1 to linear
which added a small boost and coordinate descent tuning in all
directions. See my comment below for details on each technique.

since weight-only int8 quantization is the only type of
quantization that accumulates in the native dtype, when you do the
initial matmul before the rescale, you can easily overflow. Using bf16
or fp32 fixes it. scripts altered to use bf16 instead of fp16, I don't
see a significant speed difference between them.

also added tags and enabled other types of quantization.

Test Plan: sh run.sh

Reviewers:

Subscribers:

Tasks:

Tags:
Tasks:

Tags:

### CODE REVIEW

**Comment by cpuhrsch** [comment]
nit: Don't commit the .pyc files :)

**Comment by sayakpaul** [comment]
Thanks for your PR, @HDCharles. I am away and will be back tomorrow. 

Let me know when it's ready for a review.

**Comment by HDCharles** [comment]
> Thanks for your PR, @HDCharles. I am away and will be back tomorrow.
> 
> Let me know when it's ready for a review.

its ready, though many changes were to ease my investigation and not necessarily for merge, let me know your thoughts

**Comment by HDCharles** [comment]
results of my experiments, i started at the bottom, largely ignored the other techniques due to lack of kernels for these shapes and gradually altered settings based on the tags. Final result is around 10% speedup. (scroll all the way to the right for most useful info)

| pipeline_cls              | ckpt_id                                  | bf16 | sdpa | fused_qkv_projections | upcast_vae | batch_size | num_inference_steps | compile_unet | compile_vae | compile_mode | change_comp_config | do_quant       | time (secs) | memory (gbs) | actual_gpu_memory (gbs) | tag                       |
|---------------------------|------------------------------------------|------|------|-----------------------|------------|------------|---------------------|--------------|-------------|--------------|--------------------|----------------|-------------|--------------|-------------------------|---------------------------|
| StableDiffusionXLPipeline | stabilityai/stable-diffusion-xl-base-1.0 | True | True | True                  | False      | 1          | 30                  | True         | True        | max-autotune | True               |                | 2.775       | 9.779        | 79.154                  | conv->linear              |
| StableDiffusionXLPipeline | stabilityai/stable-diffusion-xl-base-1.0 | True | True | True                  | False      | 1          | 30                  | True         | True        | max-autotune | True               |                | 2.751       | 9.540        | 79.154                  | noepi                     |
| StableDiffusionXLPipeline | stabilityai/stable-diffusion-xl-base-1.0 | True | True | True                  | False      | 1          | 30                  | True         | True        | max-autotune | True               |                | 2.794       | 10.530       | 79.154                  | epi                       |
| StableDiffusionXLPipeline | stabilityai/stable-diffusion-xl-base-1.0 | True | True | True                  | False      | 1          | 30                  | True         | True        | max-autotune | True               |                | 2.756       | 10.529       | 79.154                  | all_dir_fullgraph_new_api |
| StableDiffusionXLPipeline | stabilityai/stable-diffusion-xl-base-1.0 | True | True | True                  | False      | 1          | 30                  | True         | True        | max-autotune | True               | int8dynamic    | 2.489       | 7.839        | 79.154                  | conv->linear select       |
| StableDiffusionXLPipeline | stabilityai/stable-diffusion-xl-base-1.0 | True | True | True                  | False      | 1          | 30                  | True         | True        | max-autotune | True               | int8dynamic    | 2.542       | 7.842        | 79.154                  | conv->linear              |
| StableDiffusionXLPipeline | stabilityai/stable-diffusion-xl-base-1.0 | True | True | True                  | False      | 1          | 30                  | True         | True        | max-autotune | True               | int8dynamic    | 2.512       | 7.887        | 79.154                  | all_dir_fullgraph_new_api |
| StableDiffusionXLPipeline | stabilityai/stable-diffusion-xl-base-1.0 | True | True | True                  | False      | 1          | 30                  | True         | True        | max-autotune | True               | int8dynamic    | 2.513       | 7.879        | 79.154                  | noepi                     |
| StableDiffusionXLPipeline | stabilityai/stable-diffusion-xl-base-1.0 | True | True | True                  | False      | 1          | 30                  | True         | True        | max-autotune | True               | int8dynamic    | 2.639       | 7.879        | 79.154                  | epi                       |
| StableDiffusionXLPipeline | stabilityai/stable-diffusion-xl-base-1.0 | True | True | True                  | False      | 1          | 30                  | True         | True        | max-autotune | True               | int8weightonly | 3.084       | 7.838        | 79.154                  | noepi                     |
| StableDiffusionXLPipeline | stabilityai/stable-diffusion-xl-base-1.0 | True | True | True                  | False      | 1          | 30                  | True         | True        | max-autotune | True               | int4weightonly | 22.832      | 7.907        | 79.154                  | noepi                     |

**Comment by sayakpaul** [comment]
Looks pretty good to me. I am happy to merge it without the CSV file. I will also run a benchmark on our 400W A100 (with clock set). Let me know.

File: utils/pipeline_utils.py [hunk 1]
```diff
@@ -59,36 +47,56 @@ def load_pipeline(args):
     if args.compile_unet:
         pipe.unet.to(memory_format=torch.channels_last)
         print("Compile UNet")
-
+        swap_conv2d_1x1_to_linear(pipe.unet)
```
Comment (sayakpaul) [inline utils/pipeline_utils.py:50]: How much does it help? Should we make it configurable?

**Review by sayakpaul — APPROVED** [review]
Just one comment.

File: utils/pipeline_utils.py (hunk 1 above)
Comment (HDCharles) [reply utils/pipeline_utils.py:50]: not a lot to be honest, the shapes looked like they would be good to apply quantization to but it seems the improvements are relatively minor. Note i've altered the api's so they can be called on individual modules if needed so you could call that on individual conv modules we'd like to target.

Comment (HDCharles) [reply utils/pipeline_utils.py:50]: also,the code has torch._inductor.config.conv_1x1_as_mm = True so its already doing this, but just at a point after we apply quantization. Its unclear if the two different ways of doing conv1x1->linear have different perf on a shape to shape basis but overall we can see that the perf impact is relatively minor from [perf results](https://github.com/huggingface/sdxl-fast/pull/2#issuecomment-1855383345)

**Pull request merged**